docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+scenario we were not even able to connect more than 120 devices to two LNSs, while in a single LNS
+topology we could scale up to 900 devices without issues.
+
+## Out of scope
+
+- Deduplication strategies Mark and None: these strategies rely on multiple LNSs sending message.
+Potentially we could consider other workarounds for the IoT Hub limitation of a single connection
+per device but we find it acceptable for the Mark and None strategies to not be as scalable as the
+Drop strategy and will only document this limitation for potential users to be aware of.
+
+- LNS performs operations on behalf of a device/sensor and a concentrator/station. However since a
+concentrator can be connected to at most one LNS, there is no ping-pong happening with operations on
+stations.
+
+- Class C downstream messages using Direct Method:
+  - When we send a downstream message via the Function, the Function is responsible for choosing the
+    prererred gateway based on its existing knowledge of the preferred gateway. Due to that we can
+    be sure that we have the active connection and can therefore send downstream messages.
+  - Messages sent from the portal could result in a connection switch if they target a LNS other
+than the preferred one but we consider this as out of scope as we can not guard against this case of
+misuse.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming leaf devices (that potentially become out-of-range from an LNS) are kept in scope.
+
+### Problematic IoT Hub operations on behalf of edge devices
+
+- Background tasks
+  - Periodically we refresh the LoRaDeviceCache, which results in device twin reads that could
+    switch the connection -> see [handling of background tasks section](#handling-of-background-tasks)
+- Message flows
+  - Join -> see [handling of Join requests section](#Handling-of-Join-requests)
+  - Data:
+    - if the device is not in LoRaDeviceCache, we fetch the device twin -> see [main data flow section](#main-data-message-flow)
+    - assuming we have the device twin (in the cache or fetched) in the main data flow we send upstream, downstream and write the new twin -> see [main data flow section](#main-data-message-flow)
+    - if a frame counter reset happened, we update the twin immediately -> see [handling of resets section](#handling-of-device-resets)
+
+Version, LNS discovery and CUPS update endpoints are not affected by this issue.
+
+## Possible solution
+
+The main idea is to give the current connection holder (as indicated from the Function), the edge to
+continue processing messages for this device. The performance of that gateway will not be impacted.
+
+A local (per LNS) in-memory dictionary will be used for that purpose to map DevEuis to a flag
+indicating whether the LNS is the preferred gateway or not.
```

Comment:
I would re-write this to broader strokes and not implementation details (use of a dictionary):
```suggestion
The main idea is to give the current connection owner (as indicated from the Function), the edge to
continue processing messages for this device. The performance of that gateway will not be impacted.

If the current connection owner will keep the connection open. Any other network servers receiving messages, will not maintain an active connection to IoT Hub.

If the owning network server stops responding or gets out of reach, the ownership is transferred to the next winning network server.

```